src/psyclone/psyir/symbols/datatypes.py

```diff
@@ -362,63 +393,105 @@ def _validate_shape(self, extents):
         # dependencies.
         # pylint: disable=import-outside-toplevel
         from psyclone.psyir.nodes import DataNode, Reference
+
+        def _validate_data_node(dim_node):
+            '''
+            Checks that the supplied DataNode is valid as part of an
+            array-shape specification.
+
+            :param dim_node: the DataNode to check.
+            :type dim_node: :py:class:`psyclone.psyir.nodes.DataNode`
+
+            :raises TypeError: if the DataNode is not valid in this context.
+
+            '''
+            # When issue #685 is addressed then check that the
+            # datatype returned is an int (or is unknown). For the
+            # moment, just check that if the DataNode is a
+            # Reference then the associated symbol is a scalar
+            # integer or is unknown.
+            if isinstance(dim_node, Reference):
+                # Check the DataSymbol instance is a scalar
+                # integer or is unknown
+                symbol = dim_node.symbol
+                if not ((symbol.is_scalar and symbol.datatype.intrinsic ==
+                         ScalarType.Intrinsic.INTEGER) or
+                        isinstance(symbol.datatype,
+                                   (UnknownFortranType, DeferredType))):
+                    raise TypeError(
+                        "If a datasymbol is used as a dimension "
+                        "declaration then it should be a scalar "
+                        "integer or an unknown type, but '{0}' is a "
```

Comment:
Also mention DeferredType and capitalise DataSymbol and UnknownType.

Reply:
Done.